### PR TITLE
[IMP] *: Print GS1 datamatrices + improve labels

### PR DESCRIPTION
**Pull request by ticodoo**
This PR does a few things:

GS1 / Datamatrix (DM) related parts: 
- Adds in ability to print DM barcode (note: still requires additional package installation to work correctly)
- Adds UoM GS1 rules + yard UoM (note: some Odoo UoMs are not supported by GS1 and will not print in GS1 DMs)
- Print product GS1 barcodes in "Package Barcode with Content" report (note: only when product barcode is EAN compliant + prints additional info when applicable in this case)
- Include product UoM in DM in "Package Barcode with Content" report when valid product GS1 (note: number of digits, including precision digits, are limited in GS1 and will not print in GS1 DM if too long) 
- Add option to print GS1 DMs for Lot/SN labels (note: will only include product in GS1 barcode if EAN compliant)
- Print GS1 SSCC DM for packages when package name is a valid SSCC number (note: for reports: Package with Content, Package Barcode (PDF), and Package Barcode (ZPL))

Label improvements related parts:
- Make Lot/SN label PDF reports match ZPL layout and also follow 4x12 format (same as product label 4x12 layout)
- Allow printing "Package Barcode with Content" directly from picking
- Make MRP "Finish Product Labels" (PDF) print in standard 4x12 layout (previously was a non-standard format)
- Fix MRP label (ZPL) so that a valid zpl is always created regardless of barcode/tracking status
- Add new wizard to print a pickings' SN/Lot labels (via existing "Print Label" button)

See individual commits for additional details for each part's implementations.

ENT PR: odoo/enterprise#23323
Upgrade PR: odoo/upgrade#3202

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ticodoo** [comment]
Note that **requirements.txt change will not be merged** in end due to pylibdmtx package not being available on latest Ubuntu LTS. For now we leave it in during testing.

**Comment by amoyaux** [comment]
@ticodoo Okay but what do we do if we don't add it to requirement? You skip the feature if the package is not install with a logger warning?

File: addons/stock/report/product_templates.xml [hunk 1]
```diff
@@ -38,17 +38,28 @@
...
 ^FO100,50
-^A0N,44,33^FD<t t-esc="lot.product_id.display_name"/>^FS
+^A0N,44,33^FD<t t-out="lot.product_id.display_name"/>^FS
 ^FO100,100
-^A0N,44,33^FDLN/SN: <t t-esc="lot.name"/>^FS
+^A0N,44,33^FDLN/SN: <t t-out="lot.name"/>^FS
+                    <t t-if="env.user.has_group('stock.group_stock_lot_print_gs1') and env['ir.actions.report'].get_datamatrix_state() != 'broken'">
+                        <t t-if="lot.product_id.valid_ean" t-set="final_barcode" t-value="'01' + '0' * (14 - len(lot.product_id.barcode)) + lot.product_id.barcode"/>
+                        <t t-if="lot.product_id.tracking == 'lot'" t-set="final_barcode" t-value="(final_barcode or '') + '10' + '0' * (20 - len(lot.name)) + lot.name"/>
+                        <t t-elif="lot.product_id.tracking == 'serial'" name="datamatrix_sn" t-set="final_barcode" t-value="(final_barcode or '') + '21' + '0' * (20 - len(lot.name)) + lot.name"/>
```
Comment (svs-odoo) [inline addons/stock/report/product_templates.xml:53]: I guess the `'0' * (20 - len(lot.name))` is done to avoid to put the FNC1 into the barcode ?
The problem with this solution is the serial/lot number isn't the same as `'lot-0001'` =/= `'000000000000lot-0001'`.

Padding with 0 for the product's barcode works because the product's barcode have to be 14 digits, so if your product's barcode is `'12345670'` , we will check for `'0100000012345670'`, which is not the case for variable length AI parts.

I suggest to always place the lot/serial part of the barcode at the end so we don't have to end it by the FNC1 or fix its length, but it is a little tricky with the override...

**Comment by ticodoo** [comment]
> @ticodoo Okay but what do we do if we don't add it to requirement? You skip the feature if the package is not install with a logger warning?

Correct. There's a warning in the setting description for the `group_stock_lot_print_gs1` about the extra needed software (it should maybe be worded better or documented). In the case when packages are supposed to print a datamatrix, there's nothing indicating to the user that a datamatrix is supposed to print so they don't need a warning. Also confusing, ZPL has a built in datamatrix writer so that will always work.

**Review by amoyaux — APPROVED** [review]
Globally it's correct. But IMO it's becoming a mess with all the different templates and wizards. It would be great to create new layout under web.basic_layout with predefined nCols and nRows. Also we should add a new scss file with class associated to predefined layout with specific div size and barcode size

File: addons/stock/models/stock_picking.py [hunk 2]
```diff
@@ -1496,6 +1496,19 @@ def action_open_label_layout(self):
                 'default_picking_quantity': 'picking'},
         }
 
+    def action_open_label_type(self):
+        if self.user_has_groups('stock.group_production_lot') and self.move_line_ids.lot_id:
```
Comment (amoyaux) [inline addons/stock/models/stock_picking.py:1500]: Should also work in receipt before validate with lot_name IMO

Comment (amoyaux) [reply addons/stock/models/stock_picking.py:1500]: We can also not support it but it's strange to have different behaviors before/after action_done

Comment (ticodoo) [reply addons/stock/models/stock_picking.py:1500]: We can maybe add it as a follow on task. Right now this uses only lot_id so we can re-use the existing lot/sn report actions (i.e. we pass the lot_id.ids). If we want to support lot_name also then we will have to create some sort of custom report that can handle both lot_ids + the lot_names while also still making them all print in the same 4x12 page together....

**Review by ticodoo — COMMENTED** [review]
Outstanding things to do in a followup task:
- see if UoMs not supported by GS1 can be auto-converted so they (and their qty) aren't randomly left off GS1 barcodes
- make generic 4x12 label report template (or multiple generic nCol x nRow templates) that can be used by lots/sn, mrp_production, maybe products? 
- see if additional changes should/can be made to barcode_nomenclature/barcode_rule for more flexible barcode creation (i.e. so users can edit rules and still print barcodes that match the rule so it's not hardcoded to GS1 standards)

Also note, `pack_date` is used instead of `packaging_date` to match https://github.com/odoo/odoo/pull/82883

**Comment by ticodoo** [comment]
Also note that `pylibdmtx` has been removed from the requirements.txt, but runbot has been updated to have package installed (+ its libdmtx package dependency).

TLDR: datamatrix now prints in runbot for this branch only.

**Comment by d-fence** [comment]
override=ci/style
The unused import is used to check the lib availability

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Comment by Xavier-Do** [comment]
> Note that **requirements.txt change will not be merged** in end due to pylibdmtx package not being available on latest Ubuntu LTS. For now we leave it in during testing.

It looks like pylibdmtx was not available in ubuntu 20.04
https://packages.ubuntu.com/search?keywords=python3-pylibdmtx&searchon=names&suite=jammy&section=all
https://packages.ubuntu.com/search?suite=focal&searchon=names&keywords=python3-pylibdmtx
